### PR TITLE
[old] Feature/c light spot 

### PR DESCRIPTION
**Pull request by Takundei**
Added SpotLight.

Rendered Image :
Demo SpotLight

### CODE REVIEW

File: demos/CMakeLists.txt [hunk 1]
```diff
@@ -52,5 +52,21 @@ set_target_properties(Demo_CornellBox PROPERTIES FOLDER "Demos")
...
+set_target_properties(Demo_SpotLight PROPERTIES FOLDER "Demos")
+
+# Properties->Linker->Input->Additional Dependencies
+target_link_libraries(Demo_SpotLight ${OpenCV_LIBS} ${CORE_LIB})
+# --------
+#install
+install(TARGETS Demo_SpotLight RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
+
 #install
 install(TARGETS Demo_CornellBox RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
```
Comment (ereator) [inline demos/CMakeLists.txt:72]: It looks like this line belongs to the previous block (DEMO CORNELL BOX)

File: modules/core/LightSpot.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include "ILight.h"
+#include <corecrt_math_defines.h>
```
Comment (ereator) [inline modules/core/LightSpot.h:4]: Interesting :-)
Perhaps it makes sense to make use of it for the whole library globally. I will take care about this.

File: modules/core/LightSpot.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		DllExport virtual std::optional<Vec3f> illuminate(Ray& ray) override
+		{
+			// ray towards spot light position
+			ray.dir = ray.org - m_position;
+			ray.t = norm(ray.dir);
+			ray.dir = normalize(ray.dir);
+			float angle_light_ray = ray.dir.dot(m_direction);
+			float attenuation = 1 / (ray.t * ray.t);
+			float scale;
+			if (angle_light_ray > m_angle) {
```
Comment (ereator) [inline modules/core/LightSpot.h:38]: Shouldn't it be
`if (angle_light_ray < m_angle) {` ?

File: modules/core/LightSpot.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+				// Light Intensity reduces away from m_direction
+				scale = (1.0 - (1.0 - angle_light_ray) * 1.0 / (1.0 - m_angle));
+			}
+			else {
+				scale = 0;
```
Comment (ereator) [inline modules/core/LightSpot.h:43]: Here we can return `std::nullopt` - this will reduce computations in shaders

**Review by ereator — CHANGES_REQUESTED** [review]
One more thing - there are some files, for example Demo Cornell Box.cpp, which were changed (i think tabs were replaced with spaces or vise versa). But changing these files does not belong to the pull request in my opinion. If merge now, that in the history of that file will be a string “Add SpotLight”, but logically these files should not be affected.
I suggest to remove them from the pull request. And if those changes are still needed, add them to another pull request (e.g. “formatting code” or smth like that)

File: demos/CMakeLists.txt (hunk 1 above)
Comment (Takundei) [reply demos/CMakeLists.txt:72]: Fixed

File: modules/core/LightSpot.h (hunk 2 above)
Comment (Takundei) [reply modules/core/LightSpot.h:4]: Ok

File: modules/core/LightSpot.h (hunk 3 above)
Comment (Takundei) [reply modules/core/LightSpot.h:38]: These variables are  \cos of the respective angles, so the code would need to be : 
`if (acos(angle_light_ray) <  acos(m_angle)) {`

which version would you prefer ?

File: modules/core/LightSpot.h (hunk 4 above)
Comment (Takundei) [reply modules/core/LightSpot.h:43]: fixed

**Pull request closed without merging**